### ISSUE TITLE
[OHPC 2.1] Urgent need srun to run with openmpi

### ISSUE THREAD

**Issue by vyscenkoh** [openhpc/ohpc#1634]
In OpenHPC v2.1, whenever srun is used with openmpi, "OMPI was not built with SLURM's PMI" support error is returned.

I notice some discussion already mention openmpi do not works with slurm because of pmix.
https://github.com/openhpc/ohpc/issues/1273
https://github.com/openhpc/ohpc/issues/1320
https://github.com/stackhpc/ansible_collection_slurm_openstack_tools/pull/32

However, I am transitioning from ohpc2.0 to 2.1, I really need srun to work because my application still rely on it. 
Is there any work around? Urgent.. Please help...

Details:
srun --mpi=list > pmix,pmi2,pmix_v3
ompi_info | grep -i pmix > flux, isolated, pmix3x
yum list | grep pmix > pmix3.1.5....
Recompile / Reinstall slurm is not an option, originate from Qlustar almalinux 8.6 image
slurm 21.08.8.2
openmpi 4.1.1
hwloc2.5
libfabric1.18
ucx1.11.2
gcc9.4

**Comment by adrianreber** [comment]
@mghpcsim you mentioned somewhere you have some documentation about openhpc slurm and pmix. Can you help here?

**Comment by ohpcsim** [comment]
Here are instructions for rebuilding the relevant RPMs to re-enable PMIx support.

This was done on a fresh Rocky 8 install. It can be done on any node and then the resulting RPMs can be put into a local repository.
 
dnf -y install epel-release
dnf -y config-manager --set-enabled powertools
dnf -y install screen wget vim-enhanced 
 
dnf -y update && reboot
 
 
dnf -y install http://repos.openhpc.community/OpenHPC/2/EL_8/x86_64/ohpc-release-2-1.el8.x86_64.rpm
dnf -y install ohpc-base
dnf -y install ohpc-slurm-server
dnf -y install ohpc-autotools
dnf -y install hwloc-ohpc
dnf -y install gnu9-compilers-ohpc
dnf -y install openmpi4-gnu9-ohpc
dnf -y install lmod-defaults-gnu9-openmpi4-ohpc
 
log out and log back in
 
dnf -y install rpm-build yum-utils
yum-builddep openmpi4-gnu9-ohpc
yum-builddep pmix-ohpc
yum-builddep ohpc-slurm-server slurm-ohpc slurm-slurmd-ohpc slurm-slurmctld-ohpc
 
 
mkdir src && cd src
yumdownloader --source pmix-ohpc
yumdownloader --source openmpi4-gnu9-ohpc
yumdownloader --source slurm-slurmd
rpm -i *rpm
 
cd /root/rpmbuild/SOURCES
wget https://github.com/openpmix/openpmix/releases/download/v3.2.2/pmix-3.2.2.tar.bz2
 
 
cd /root/rpmbuild/SPECS
edit pmix.spec
 
Add
%undefine _annotated_build
Under the line that loads OHPC Macros (top of file)
Set the version to 3.2.2
And increment the major number of the Release (so it’ll be seen as an update to dnf)
 
rpmbuild -bb pmix.spec
cd /root/rpmbuild/RPMS
dnf install pmix*rpm
 
module load pmix
 
cd /root/rpmbuild/SPECS
edit openmpi.spec
and change the line
%{!?with_pmix: %define with_pmix 0}
To
%{!?with_pmix: %define with_pmix 1}
And increment the major number of the Release
 
rpmbuild -bb openmpi.spec
cd /root/rpmbuild/RPMS
dnf install openmpi*rpm
 
cd /root/rpmbuild/SPECS
edit slurm.spec
 
Add
%undefine _annotated_build
Under the line that loads OHPC Macros (top of file)
And increment the major number of the Release
 
rpmbuild -bb slurm.spec
 
cd /root/rpmbuild/RPMS
dnf install slurm*rpm

**Comment by vyscenkoh** [comment]
Without reinstalling slurm is impossible to workaround it? 
Please note that,
Recompile / Reinstall slurm is not an option, originate from Qlustar almalinux 8.6 image
OHPC2.1 is originate from Qlustar, an intranet HPC Management environment.

**Comment by ohpcsim** [comment]
Unfortunately, PMIx support for SLURM is done at compile time; so yes, you need to recompile / reinstall. I'm not sure why recompile / reinstall is not an option. I'd suggest reaching out to Qlustar and asking for guidance. They should be able to help you with this if you have a support contract. Pushing updates should be fairly easy.

**Comment by sjpb** [comment]
Oh so this is interesting, OpenHPC slurm is built with `pmi2` support but OpenHPC openmpi supports pmi(?) and pmix:

```
[rocky@demo-login-0 pingpong]$ srun --mpi=list
MPI plugin types are...
        cray_shasta
        none
        pmi2
[rocky@demo-login-0 pingpong]$ module load gnu12/12.2.0 openmpi4/4.1.4
[rocky@demo-login-0 pingpong]$ ompi_info | grep -i pmi
                MCA pmix: flux (MCA v2.1.0, API v2.0.0, Component v4.1.4)
                MCA pmix: isolated (MCA v2.1.0, API v2.0.0, Component v4.1.4)
                MCA pmix: pmix3x (MCA v2.1.0, API v2.0.0, Component v4.1.4)
                 MCA ess: pmi (MCA v2.1.0, API v3.0.0, Component v4.1.4)
```

(I don't know why I've got a different srun mpi list than @vyscenkoh , maybe it's changed - I'm on latest OpenHPC).

This seems un-ideal. Is it something which could get changed in a later release, or are there reasons for the mis-match? Seems like having srun work with OpenHPC MPI's would be a win?

**Comment by sjpb** [comment]
@mghpcsim thanks for your recipe above for rebuilding the rpms - am I right that for slurm itself, that's the spec file from an EPEL slurm, not OpenHPC? Otherwise presumably it'd be `yumdownloader --source slurm-slurmd-ohpc` not `yumdownloader --source slurm-slurmd` as given?

If so, is there a reason why please? I'd have assumed using the OpenHPC one would be necessary/preferable to get a set of slurm packages compatible with the rest of the OpenHPC packages.

**Comment by whophil** [comment]
I am just stumbling across this now. I suppose this is not yet fixed in latest OHPC?

I have the following packages installed
```
openmpi4-intel-ohpc.x86_64                                         4.1.4-6.2.ohpc.2.6                                      @OpenHPC-updates
ohpc-slurm-server.x86_64                                           2.6-7.1.ohpc.2.6                                        @OpenHPC-updates
pdsh-mod-slurm-ohpc.x86_64                                         2.34-9.1.ohpc.2.6                                       @OpenHPC-updates
slurm-contribs-ohpc.x86_64                                         22.05.2-14.1.ohpc.2.6                                   @OpenHPC-updates
slurm-devel-ohpc.x86_64                                            22.05.2-14.1.ohpc.2.6                                   @OpenHPC-updates
slurm-example-configs-ohpc.x86_64                                  22.05.2-14.1.ohpc.2.6                                   @OpenHPC-updates
slurm-libpmi-ohpc.x86_64                                           22.05.2-14.1.ohpc.2.6                                   @OpenHPC-updates
slurm-ohpc.x86_64                                                  22.05.2-14.1.ohpc.2.6                                   @OpenHPC-updates
slurm-perlapi-ohpc.x86_64                                          22.05.2-14.1.ohpc.2.6                                   @OpenHPC-updates
slurm-slurmctld-ohpc.x86_64                                        22.05.2-14.1.ohpc.2.6                                   @OpenHPC-updates
slurm-slurmdbd-ohpc.x86_64                                         22.05.2-14.1.ohpc.2.6                                   @OpenHPC-updates
```
And get the following output:

```
$ srun --mpi=list
MPI plugin types are...
	none
	cray_shasta
	pmi2
```

```
$ ompi_info | grep -i pmi
                MCA pmix: flux (MCA v2.1.0, API v2.0.0, Component v4.1.4)
                MCA pmix: pmix3x (MCA v2.1.0, API v2.0.0, Component v4.1.4)
                MCA pmix: isolated (MCA v2.1.0, API v2.0.0, Component v4.1.4)
                 MCA ess: pmi (MCA v2.1.0, API v3.0.0, Component v4.1.4)
```

**Comment by iGeorgeX** [comment]
Is the missing pmix support still an issue in the 2.7 version?

**Comment by adrianreber** [comment]
> Is the missing pmix support still an issue in the 2.7 version?

Yes. Nothing has changed with regards to pmix and Open MPI and slurm. https://github.com/openhpc/ohpc/blob/2.x/components/rms/slurm/SPECS/slurm.spec#L13

We enabled it in the 3.x branch.

**Comment by iGeorgeX** [comment]
But it is fixed in 2.8? [Slurm doesn't use pmix in ohpc v2.1 #1320](https://github.com/openhpc/ohpc/issues/1320)

**Comment by sjpb** [comment]
Should be, your linked issue got closed with the release of 2.8: https://github.com/openhpc/ohpc/issues/1320#issuecomment-2077170091
